### ISSUE TITLE
Conditional conversation nodes

### ISSUE THREAD

**Issue by SolraBizna** [endless-sky/endless-sky#5439]
**Is your feature request related to a problem? Please describe.**

I'm working on a plugin with a very non-linear, highly interdependent mission tree, to the point that it's more like a mission web. Very often, I'll have a conversation node where I want to offer six or more choices to the player, except that many of the choices require and/or are exclusive with things that may or may not have happened in previous missions. The usual solution for conditional `choice` nodes in Endless Sky is ... inadequate at this level.

**Related Issue Links**

(A quick search didn't find anybody else requesting this.)

**Describe the solution you'd like**

Synthesizing a very silly example:

```sh
conversation
    `"Hello!" the man sitting next to you at the bar says. "You're Captain <last>, right?"`
    branch "don't recognize him"
        not "got told about the renegade deputy"
    `    You recognize his face from the wanted bulletin that Captain Snafu sent you. This is the deputy that let the criminals who stole the Boundless Frontiers source code escape.`
        to display
            not "learned the truth about Boundless Frontiers LLC"
    `    You recognize his face from the wanted bulletin that Captain Snafu sent you. This is the deputy that helped those Remnant scientists steal the Boundless Frontiers source code so they could work on a cure.`
        to display
            has "learned the truth about Boundless Frontiers LLC"
    label "don't recognize him"
    `    You have to be careful what you say. Potato Pete warned you that this bar has listening devices under the stools.`
        to display
            has "know about the listening device"
    choice
        `    "Uh... hello."`
            goto "hello" # (clearer than just falling through)
        `    (Ignore him.)`
            defer
        `    (Try miming to him that there are listening devices under the stools.)`
            to display
                has "know about the listening device"
            goto "get arrested for public miming"
        `    (Try answering him in Remnant sign language.)`
            to display
                has "event: remnant: sign studies complete"
                or
                    has "know about the listening device"
                    has "learned the truth about Boundless Frontiers LLC"
            goto "get away with helping the deputy"
        `    (Shoot him.)`
            to display
                has "got told about the renegade deputy"
            goto "maybe shoot the deputy" # ...but NOT the sheriff?
        `    "You're under arrest!"`
            to display
                has "got told about the renegade deputy"
            goto "deputy runs away"
        `    "Hey, keep your voice down. You're a wanted man."`
            to display
                has "got told about the renegade deputy"
            goto "get caught helping the deputy"
```

The new `to display` node can go as a child of either a normal text node or the child of a `choice` node. If the condition is not met, it is as if the node was erased. So, in the above example:

```sh
        `    (Try miming to him that there are listening devices under the stools.)`
            to display
                has "know about the listening device"
            goto "get arrested for public miming"
```

If the "know about the listening device" condition is not set, it is as though the "miming" choice wasn't present in the list.

**Describe alternatives you've considered**

In the above example, choosing between the two "you recognize his face" parts can be handled with an additional `branch`, and only be a little clunkier. But even with just three binary conditions ("know about the listening device", "know remnant sign language and have either of two reasons to try it here" and "saw the wanted poster") the branch solution would require me to manually create eight different `choice` nodes and make a three-level branch mess to choose the right one. And then, if I want to add another choice, or I realize a particular case I missed, I'm really up a creek. In the unusually egregious case of the real mission I was just working on, there were four (or five? not sure yet) conditions, which would require me to manually make sixteen (or thirty-two!) choice nodes.

**Additional context**

~~I made a very brief attempt to add support for this to `Conversation` myself, but my first pass over the code did not yield sufficient understanding to make the necessary changes. I'll make a more serious attempt later. If that fails, I'll take the truly disgusting path of making [yet](https://github.com/SolraBizna/hubris) [another](https://github.com/SolraBizna/elko) thin DSL for this.~~

~~(And, I'm fully aware of the logistical nightmare that I'm unleashing on myself, by creating so many different paths both through individual conversations and through the mission tree as a whole. I'm a NetHack veteran, so... believe me, I know.)~~

PR #5440 implements this. It hasn't merged yet because this is pretty core stuff and we want to be sure we get it right.

Edited 2022-11-17: linked PR, updated example syntax

**Comment by InfiniteDonuts** [comment]
This *would* be nice, really. I would like to be able to handle such things as events and conditions inside conversations.

Also please, *please*, sneak this conversation into a PR -- I mean, that's not very professional of you; this is a serious development community, not a comedy club.

**Comment by SolraBizna** [comment]
I actually made a second attempt over the next few days, and a working implementation of the feature is now at #5440 waiting for more testing. I tested it myself with the early versions of my horrifying mission web to good effect, and also played through 80% of the existing story content without any issues. And then I got distracted.

In fact, while playing through existing story content, I found several places in existing missions where this feature would streamline things. I should go through and add those to the PR. If only I had more free time...
